src/usr/local/www/firewall_rules_edit.php

```diff
@@ -311,27 +311,27 @@ function is_aoadv_used($rule_config) {
 						continue;
 					}
 					if (($_POST['ipprotocol'] == "inet46") && ($_POST['ipprotocol'] != $family)) {
-						$input_errors[] = gettext("A gateway can not be assigned to a rule that applies to IPv4 and IPv6");
+						$input_errors[] = gettext("Gateways can not be assigned in a rule that applies to both IPv4 and IPv6.");
 					}
 					if (($_POST['ipprotocol'] == "inet6") && ($_POST['ipprotocol'] != $family)) {
-						$input_errors[] = gettext("An IPv4 gateway group can not be assigned on an IPv6 Address Family rule");
+						$input_errors[] = gettext("An IPv4 gateway group can not be assigned in IPv6 rules.");
 					}
 					if (($_POST['ipprotocol'] == "inet") && ($_POST['ipprotocol'] != $family)) {
-						$input_errors[] = gettext("An IPv6 gateway group can not be assigned on an IPv4 Address Family rule");
+						$input_errors[] = gettext("An IPv6 gateway group can not be assigned in IPv4 rules.");
 					}
 				}
 			}
 		}
-	}
-	if (($_POST['ipprotocol'] <> "") && ($_POST['gateway'] <> "") && (is_ipaddr(lookup_gateway_ip_by_name($_POST['gateway'])))) {
-		if (($_POST['ipprotocol'] == "inet46") && ($_POST['gateway'] <> "")) {
-			$input_errors[] = gettext("A gateway can not be assigned to a rule that applies to IPv4 and IPv6");
-		}
-		if (($_POST['ipprotocol'] == "inet6") && (!is_ipaddrv6(lookup_gateway_ip_by_name($_POST['gateway'])))) {
-			$input_errors[] = gettext("An IPv4 Gateway can not be assigned to an IPv6 Filter rule");
-		}
-		if (($_POST['ipprotocol'] == "inet") && (!is_ipaddrv4(lookup_gateway_ip_by_name($_POST['gateway'])))) {
-			$input_errors[] = gettext("An IPv6 Gateway can not be assigned to an IPv4 Filter rule");
+		if (is_ipaddr(lookup_gateway_ip_by_name($_POST['gateway']))) {
+			if (($_POST['ipprotocol'] == "inet46") && ($_POST['gateway'] <> "")) {
+				$input_errors[] = gettext("Gateways can not be assigned in a rule that applies to both IPv4 and IPv6.");
+			}
+			if (($_POST['ipprotocol'] == "inet6") && (!is_ipaddrv6(lookup_gateway_ip_by_name($_POST['gateway'])))) {
+				$input_errors[] = gettext("An IPv4 gateway can not be assigned in IPv6 rules.");
+			}
+			if (($_POST['ipprotocol'] == "inet") && (!is_ipaddrv4(lookup_gateway_ip_by_name($_POST['gateway'])))) {
+				$input_errors[] = gettext("An IPv6 gateway can not be assigned in IPv4 rules.");
+			}
 		}
```

Comment:
If $_POST['gateway'] doesn't exist, or is blank or not a valid gateway for any reason, then lookup_gateway_ip_by_name() isn't going to be able to return a valid IP address for it, and looks almost certainly like it returns false. Additionally all the is_ipaddr*() functions return a falsey value if the string isn't a valid IP (of the given type if specified). So any case where is_ipaddr*() allowed a TRUE value for the if() clause, $_POST['gateway'] would have had to be non-empty. I think that clause is really just a "belt and braces" safety check that the value passed to the other functions exists and isn't empty. But it doesn't matter because the rest does in fact look like it will handle that sort of thing correctly if not. 

That said it can be made more efficient. Let's do so.
is defined in gwlb.inc (about line 1037), and is_ipaddr*() are defined in util.inc.